File: 1580-put-boxes-into-the-warehouse-ii/1580-put-boxes-into-the-warehouse-ii.cpp
class Solution {
public:
    int maxBoxesInWarehouse(vector<int>& boxes, vector<int>& warehouse) {
        // insert box from both sides.
        int m = boxes.size(), n = warehouse.size();
        sort(boxes.begin(), boxes.end());

        int left = 0, right = n - 1, count = 0, boxIndex = m - 1;
        while (left <= right && boxIndex >= 0) {
            if (boxes[boxIndex] <= warehouse[left]) {
                ++ left;
                ++ count;
            } else if (boxes[boxIndex] <= warehouse[right]) {
                -- right;
                ++ count;
            }
            -- boxIndex;
        }
        return count;
    }
};

File: 1580-put-boxes-into-the-warehouse-ii/1580-put-boxes-into-the-warehouse-ii.java
class Solution {
    public int maxBoxesInWarehouse(int[] boxes, int[] warehouse) {
        int l = 0, r = warehouse.length - 1;
        Arrays.sort(boxes);
        int b = boxes.length - 1, res = 0;
        while (l <= r) {
            while (b >= 0 && warehouse[l] < boxes[b] && warehouse[r] < boxes[b]) b--;

            if (b == -1) return res;

            if (warehouse[l] >= boxes[b] && warehouse[r] >= boxes[b]) {
                if (warehouse[l] <= warehouse[r]) l++;
                else r--;
            } else if (warehouse[l] >= boxes[b]) {
                l++;
            } else {
                r--;
            }

            b--;
            res++;
        }
        return res;
    }
}